File: backend/app/models/tasks.py
from datetime import date, datetime, time
from typing import Literal

from pydantic import BaseModel, Field

VALID_STATUSES = ("todo", "in_progress", "done", "cancelled")
VALID_RECURRENCES = ("daily", "weekly", "monthly")


class TaskGroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    color: str = Field(default="#6366f1", max_length=20)


class TaskGroupUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    color: str | None = Field(default=None, max_length=20)
    sort_order: int | None = None


class TaskGroupResponse(BaseModel):
    id: str
    user_id: str
    name: str
    color: str
    sort_order: int
    created_at: datetime


class TaskCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    description: str | None = None
    priority: int = Field(default=0, ge=0, le=4)
    status: Literal["todo", "in_progress", "done", "cancelled"] = "todo"
    due_date: date | None = None
    due_time: time | None = None
    recurrence: Literal["daily", "weekly", "monthly"] | None = None
    group_id: str | None = None
    parent_task_id: str | None = None


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = None
    priority: int | None = Field(default=None, ge=0, le=4)
    status: Literal["todo", "in_progress", "done", "cancelled"] | None = None
    due_date: date | None = None
    due_time: time | None = None
    recurrence: Literal["daily", "weekly", "monthly"] | None = None
    group_id: str | None = None


class TaskStatusUpdate(BaseModel):
    status: Literal["todo", "in_progress", "done", "cancelled"]


class TaskResponse(BaseModel):
    id: str
    user_id: str
    group_id: str | None
    parent_task_id: str | None
    title: str
    description: str | None
    priority: int
    status: str
    due_date: date | None
    due_time: time | None
    recurrence: str | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    sort_order: int = 0
    subtask_count: int = 0
    subtask_done_count: int = 0
    labels: list["LabelResponse"] = []


class TaskDetailResponse(TaskResponse):
    subtasks: list[TaskResponse] = []
    reminders: list["ReminderResponse"] = []
    group: TaskGroupResponse | None = None


class ReminderCreate(BaseModel):
    remind_at: datetime
    type: Literal["push", "email"] = "push"


class ReminderResponse(BaseModel):
    id: str
    task_id: str
    remind_at: datetime
    type: str
    is_sent: bool
    created_at: datetime


class ReorderItem(BaseModel):
    id: str
    sort_order: int


class ReorderTasksPayload(BaseModel):
    items: list[ReorderItem] = Field(..., min_length=1)


class LabelCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    color: str = Field(default="#6366f1", max_length=20)


class LabelUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    color: str | None = Field(default=None, max_length=20)


class LabelResponse(BaseModel):
    id: str
    user_id: str
    name: str
    color: str
    created_at: datetime


class DailySummaryResponse(BaseModel):
    overdue_count: int
    today_tasks: list[TaskResponse]
    upcoming_tasks: list[TaskResponse]

File: frontend/src/types/tasks.ts
export interface TaskGroup {
  id: string;
  user_id: string;
  name: string;
  color: string;
  sort_order: number;
  created_at: string;
}

export interface Task {
  id: string;
  user_id: string;
  group_id: string | null;
  parent_task_id: string | null;
  title: string;
  description: string | null;
  priority: number;
  status: string;
  due_date: string | null;
  due_time: string | null;
  recurrence: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  sort_order: number;
  subtask_count: number;
  subtask_done_count: number;
  labels: Label[];
}

export interface Reminder {
  id: string;
  task_id: string;
  remind_at: string;
  type: string;
  is_sent: boolean;
  created_at: string;
}

export interface TaskDetail extends Task {
  subtasks: Task[];
  reminders: Reminder[];
  group: TaskGroup | null;
  labels: Label[];
}

export interface DailySummary {
  overdue_count: number;
  today_tasks: Task[];
  upcoming_tasks: Task[];
}

export interface CreateTaskPayload {
  title: string;
  description?: string;
  priority?: number;
  status?: string;
  due_date?: string;
  due_time?: string;
  recurrence?: string;
  group_id?: string;
  parent_task_id?: string;
}

export interface UpdateTaskPayload {
  title?: string;
  description?: string;
  priority?: number;
  status?: string;
  due_date?: string;
  due_time?: string;
  recurrence?: string;
  group_id?: string;
}

export interface CreateTaskGroupPayload {
  name: string;
  color?: string;
}

export interface UpdateTaskGroupPayload {
  name?: string;
  color?: string;
  sort_order?: number;
}

export interface CreateReminderPayload {
  remind_at: string;
  type?: "push" | "email";
}

export interface Label {
  id: string;
  user_id: string;
  name: string;
  color: string;
  created_at: string;
}

export interface CreateLabelPayload {
  name: string;
  color?: string;
}

export interface UpdateLabelPayload {
  name?: string;
  color?: string;
}
